source/http_stream.c

```diff
@@ -16,65 +16,65 @@
 #include "http_stream.h"
 #include "http_connection.h"
 #include "module.h"
-#include "uv_interop.h"
 
 #include <aws/http/request_response.h>
 #include <aws/io/stream.h>
 
 struct http_stream_binding {
     struct aws_http_stream *stream;
+    struct aws_allocator *allocator;
     napi_ref node_external;
-    struct aws_uv_context *uv_context;
-    struct aws_napi_callback on_complete;
-    struct aws_napi_callback on_response;
-    struct aws_napi_callback on_body;
+    napi_threadsafe_function on_complete;
+    napi_threadsafe_function on_response;
+    napi_threadsafe_function on_body;
     struct aws_http_message *response; /* used to buffer response headers/status code */
     struct aws_http_message *request;
     struct aws_input_stream *body_stream; /* stream pointing at request body */
 };
 
-static int s_on_response_params(napi_env env, napi_value *params, size_t *num_params, void *user_data) {
+static void s_on_response_call(napi_env env, napi_value on_response, void *context, void *user_data) {
+    struct http_stream_binding *binding = context;
     struct aws_http_message *response = user_data;
 
-    int32_t status_code = 0;
-    aws_http_message_get_response_status(response, &status_code);
-    if (napi_create_int32(env, status_code, &params[0])) {
-        return AWS_OP_ERR;
-    }
-
-    napi_value node_headers = NULL;
-    if (napi_create_array(env, &node_headers)) {
-        return AWS_OP_ERR;
-    }
-    params[1] = node_headers;
-
-    const size_t num_headers = aws_http_message_get_header_count(response);
-    for (size_t idx = 0; idx < num_headers; ++idx) {
-        struct aws_http_header header;
-        aws_http_message_get_header(response, &header, idx);
-
-        napi_value node_header = NULL;
-        if (napi_create_array(env, &node_header)) {
-            return AWS_OP_ERR;
+    if (env) {
+        napi_value params[2];
+        const size_t num_params = AWS_ARRAY_SIZE(params);
+
+        int32_t status_code = 0;
+        aws_http_message_get_response_status(response, &status_code);
+        AWS_NAPI_ENSURE(env, napi_create_int32(env, status_code, &params[0]));
+
+        napi_value node_headers = NULL;
+        AWS_NAPI_ENSURE(env, napi_create_array(env, &node_headers));
+        params[1] = node_headers;
+
+        const size_t num_headers = aws_http_message_get_header_count(response);
+        for (size_t idx = 0; idx < num_headers; ++idx) {
+            struct aws_http_header header;
+            aws_http_message_get_header(response, &header, idx);
+
+            napi_value node_header = NULL;
+            AWS_NAPI_ENSURE(env, napi_create_array(env, &node_header));
+
+            napi_value node_name = NULL;
+            napi_value node_value = NULL;
+            AWS_NAPI_ENSURE(
```

Comment:
AWS_NAPI_ENSURE is a bold approach to error handling

I've been wondering how this kind of thing should be handled (errors while prepping a callback) in the python CRT.
I've definitely been bit where elasticurl was failing in CI because we weren't waiting for the connection to finish shutdown (which should be fine in Python), so the on_shutdown in C tried to call on_shutdown in Python, which failed, so I AWS_FATAL_ASSERTed, which made it look like elasticurl was failing.

Is Node culture that an application should 100% close things down before exiting culture?

Reply:
This is only for use in the case where throwing an exception or something is not a viable alternative for handling a failure. In this case, we're in the node thread, about to call a node function, and we're pushing the arguments onto the local scope. We cannot throw an exception, because it is physically impossible for someone to catch it, so it would just prevent execution until our exception clearing code runs (see module.c). Therefore, the only option is to crash hard, there's no real recovery. Also, if node fails to create an int or a string, it's not going to succeed at much after that point.

If you look at the node-addon-examples github repo, this pattern is basically what they do. They assert on everything that is very unlikely to fail, and catch on things that are prone to failure. I have tried where possible to follow that example. That's why we have ```AWS_NAPI_ENSURE``` and ```AWS_NAPI_CALL```.